mathematica/mathematica.hpp

```diff
@@ -25,90 +28,146 @@ using geometry::R3Element;
 using geometry::Vector;
 using numerics::FixedVector;
 using physics::DegreesOfFreedom;
+using quantities::Amount;
+using quantities::Angle;
+using quantities::Current;
+using quantities::Exponentiation;
+using quantities::is_tuple_v;
+using quantities::Length;
+using quantities::LuminousIntensity;
+using quantities::Mass;
+using quantities::Pow;
 using quantities::Quantity;
 using quantities::Quotient;
+using quantities::Temperature;
+using quantities::Time;
+
+// A helper class for type erasure of quantities.  It may be used with the
+// functions in this file to remove the dimensions of quantities (we know that
+// Mathematica is sluggish when processing quantities).  Usage:
+//
+//   ToMathematica(... , ExpressIn(Metre, Second, Degree));
+//
+// The construction parameters must be for the base SI units.  They may be in
+// any order.  If not enough parameters are specified for complete type erasure
+// of the argument, compilation fails.  An object with no template parameters
+// may be used to indicate that type erasure should not happen.
```

Comment:
Changed the comments.

Trust me, `optional` is where I started, and that doesn't work because:
1. CTAD doesn't work in that case at the call site, and it becomes super-ugly to have to specify the template parameters as in `ExpressIn<Angle, Length>(Degree, Metre)`.
1. The check that we were given the default must be a constexpr if, to make sure that we don't try to compile the other branch of the if, which would not work because it doesn't have the units we need.

So we effectively need a "compile-time optional", here the special casing of `ExpressIn<>`.  It's not like it is going to cause bugs anyway: if you write `ExpressIn{}` explicitly, you are mildly insane.

I'll add this this code is really at the bleeding edge of what the language can do, and large amounts of godbolting were needed to come up with this solution, and check that it works with the compilers involved.

Reply:
Done, as discussed offline.